meta/autoid/autoid.go

```diff
@@ -1062,46 +1100,55 @@ func TestModifyBaseAndEndInjection(alloc Allocator, base, end int64) {
 	alloc.(*allocator).mu.Unlock()
 }
 
-// ShardIDLayout is used to calculate the bits length of different segments in auto id.
-// Generally, an auto id is consist of 3 segments: sign bit, shard bits and incremental bits.
+// ShardIDFormat is used to calculate the bit length of different segments in auto id.
+// Generally, an auto id is consist of 4 segments: sign bit, reserved bits, shard bits and incremental bits.
 // Take ``a BIGINT AUTO_INCREMENT PRIMARY KEY`` as an example, assume that the `shard_row_id_bits` = 5,
 // the layout is like
-//  | [sign_bit] (1 bit) | [shard_bits] (5 bits) | [incremental_bits] (64-1-5=58 bits) |
-// Please always use NewShardIDLayout() to instantiate.
-type ShardIDLayout struct {
+//  | [sign_bit] (1 bit) | [reserved bits] (0 bits) | [shard_bits] (5 bits) | [incremental_bits] (64-1-5=58 bits) |
+// Please always use NewShardIDFormat() to instantiate.
+type ShardIDFormat struct {
 	FieldType *types.FieldType
 	ShardBits uint64
 	// Derived fields.
-	TypeBitsLength  uint64
 	IncrementalBits uint64
-	HasSignBit      bool
 }
 
-// NewShardIDLayout create an instance of ShardIDLayout.
-func NewShardIDLayout(fieldType *types.FieldType, shardBits uint64) *ShardIDLayout {
-	typeBitsLength := uint64(mysql.DefaultLengthOfMysqlTypes[mysql.TypeLonglong] * 8)
-	incrementalBits := typeBitsLength - shardBits
+// NewShardIDFormat create an instance of ShardIDFormat.
+// RangeBits means the bit length of the sign bit + shard bits + incremental bits.
+// If RangeBits is 0, it will be calculated according to field type automatically.
+func NewShardIDFormat(fieldType *types.FieldType, shardBits, rangeBits uint64) ShardIDFormat {
+	var incrementalBits uint64
+	if rangeBits == 0 {
+		// Zero means that the range bits is not specified. We interpret it as the default length(BIGINT).
+		typeBitsLength := uint64(mysql.DefaultLengthOfMysqlTypes[mysql.TypeLonglong] * 8)
+		incrementalBits = typeBitsLength - shardBits
+	} else {
+		incrementalBits = rangeBits - shardBits
+	}
 	hasSignBit := !mysql.HasUnsignedFlag(fieldType.GetFlag())
 	if hasSignBit {
 		incrementalBits--
 	}
-	return &ShardIDLayout{
+	return ShardIDFormat{
 		FieldType:       fieldType,
 		ShardBits:       shardBits,
-		TypeBitsLength:  typeBitsLength,
 		IncrementalBits: incrementalBits,
-		HasSignBit:      hasSignBit,
 	}
 }
 
-// IncrementalBitsCapacity returns the max capacity of incremental section of the current layout.
-func (l *ShardIDLayout) IncrementalBitsCapacity() uint64 {
-	return uint64(math.Pow(2, float64(l.IncrementalBits))) - 1
+// IncrementalBitsCapacity returns the max capacity of incremental section of the current format.
+func (s *ShardIDFormat) IncrementalBitsCapacity() uint64 {
+	return uint64(math.Pow(2, float64(s.IncrementalBits))) - 1
+}
+
+// IncrementalMask returns 00..0[11..1], where [11..1] is the incremental part of the current format.
+func (s *ShardIDFormat) IncrementalMask() int64 {
```

Comment:
I think the result between `IncrementalBitsCapacity` and `IncrementalMask` are always same? Because the max `IncrementalBits` is 63.

Reply:
Sure. They are the same.